### DISCUSSION TITLE
Problem with services

### DISCUSSION THREAD

**Discussion by DarkoBiersack** [eProsima/DDS-Router#333] [Q&A] [answered]
Hello everybody, I'm trying to connect two machines running ROS2 nodes with this router. I've manage to route topics easly, but I noticed somethings odd. Services doen't work at all. My idea is that the router needs some sort of handsake at the begining, so the request of a sevice get lost. 
This happen also at the beginning with topics, in fact the first 5/10 messages are always not delivered.
Does anyone has ecountered these problems?

**Comment by jparisu** [comment, marked as answer]
Hi @DarkoBiersack ,

We are glad to know you are using the DDS Router. We really appreciate user feedback.
First of all, can you provide the version you are using? DDS Router version and ROS2 version.

I think there are 2 different issues here that are not related:

> Services doen't work at all. 

If you are using a vendor other than Fast DDS (other rmw different than `rmw_fastrtps`) it would probably not work.
This is because ROS2 services rely on a RPC over DDS protocol that is not yet standarized in ROS2. Fast DDS implements one version of the protocol, while other vendors could implement other ones (if curious you can check the spec: https://www.omg.org/spec/DDS-RPC/1.0/)

So, depends on your RMW, that will be other than Fast DDS in galactic, it may fail.
But fix is easy, just use fastrtps rmw: https://docs.ros.org/en/galactic/Installation/DDS-Implementations/Working-with-eProsima-Fast-DDS.html#switch-to-rmw-fastrtps

> This happen also at the beginning with topics, in fact the first 5/10 messages are always not delivered.

DDS middleware have several Quality of Service configurations (QoS).
It looks like you are using `VOLATILE` QoS (used by default) and thus messages that have been sent **before** matching are not received.
Check Durability QoS in this link: https://docs.ros.org/en/rolling/Concepts/About-Quality-of-Service-Settings.html#qos-policies
